### PR TITLE
Sherlock Audit fixes

### PR DESCRIPTION
**Pull request by asquare08**
### VUSD Fixes
[Issue#162](https://github.com/sherlock-audit/2023-04-hubble-exchange-judging/issues/162), [Issue#158]( https://github.com/sherlock-audit/2023-04-hubble-exchange-judging/issues/158), [Issue#153](https://github.com/sherlock-audit/2023-04-hubble-exchange-judging/issues/153)

- Make mint, withdraw nonReentrant
- mint back vusd to the user if withdraw failed
- don’t store return data of .call
- cap gasLimit of .call

### Oracle fixes
[Issue#18](https://github.com/sherlock-audit/2023-04-hubble-exchange-judging/issues/18)

- Read the [updatedAt](https://docs.chain.link/data-feeds/api-reference#latestrounddata) return value from the `Chainlink.latestRoundData()` function and verify that is not older than than specific time tolerance.

### CODE REVIEW

File: contracts/VUSD.sol [hunk 1]
```diff
@@ -71,14 +75,16 @@ contract VUSD is ERC20PresetMinterPauserUpgradeable, ReentrancyGuard, IVUSD {
...
+            i += 1;
 
-            (bool success, bytes memory data) = withdrawal.usr.call{value: withdrawal.amount}("");
+            (bool success, ) = withdrawal.usr.call{value: withdrawal.amount, gas: maxGas}("");
             if (success) {
                 reserve -= withdrawal.amount;
             } else {
-                emit WithdrawalFailed(withdrawal.usr, withdrawal.amount, data);
+                // in case of failure, mint back the vusd to the user
+                _mint(withdrawal.usr, withdrawal.amount / SCALING_FACTOR);
```
Comment (atvanguard) [inline contracts/VUSD.sol:85]: should we store a failed mapping that allows users to claim themselves?

File: contracts/NewOracle.sol [hunk 2]
```diff
@@ -29,8 +34,10 @@ contract NewOracle is IOracle, VanillaGovernable {
         if (stablePrice[underlying] != 0) {
             return stablePrice[underlying];
         }
-        (,answer,,,) = AggregatorV3Interface(chainLinkAggregatorMap[underlying]).latestRoundData();
+        uint updatedAt;
+        (,answer,, updatedAt,) = AggregatorV3Interface(chainLinkAggregatorMap[underlying].aggregator).latestRoundData();
```
Comment (atvanguard) [inline contracts/NewOracle.sol:38]: should we call `getLatestRoundData` here to avoid duplication?

Comment (asquare08) [reply contracts/NewOracle.sol:38]: yeah I thought so too but `getLatestRoundData` searches for previous price data if latest price is negative, this is ok for twap calculation but we don't want that in `getUnderlyingPrice`.

File: contracts/VUSD.sol (hunk 1 above)
Comment (asquare08) [reply contracts/VUSD.sol]: changed it

File: contracts/VUSD.sol [hunk 3]
```diff
@@ -71,19 +78,31 @@ contract VUSD is ERC20PresetMinterPauserUpgradeable, ReentrancyGuard, IVUSD {
...
 
+    /**
+     * @notice Rescue failed withdrawal. Mints back the vusd to the user
+    */
+    function rescueFailedWithdrawal() external nonReentrant {
+        address user = _msgSender();
+        uint amount = failedWithdrawals[user];
+        require(amount > 0, "VUSD: No failed withdrawal");
+        failedWithdrawals[user] = 0;
+        _mint(user, amount);
```
Comment (atvanguard) [inline contracts/VUSD.sol:103]: why mint? we can just transfer gas to them because their withdrawal should already have been processed.

**Review by atvanguard — APPROVED** [review]

**Pull request merged**